### ISSUE TITLE
[feat] Model sharding + data parallel + offload

### ISSUE THREAD

**Issue by blefaudeux** [facebookresearch/fairscale#277]
## 🚀 Feature
Inspired by https://github.com/pytorch/pytorch/issues/42849, propose a model sharding + optimizer state sharding + gradient sharding, with or without offloading to CPU

## Motivation
If the model is too big to fit in a given GPU memory, then the current OSS+ShardedDDP is not helping. We already propose Pipe for this, but it could be interesting to investigate interest for a different approach, where the model is either sharded in between the ranks (but data parallel is maintained, in that all ranks contribute to FW and BW), at the cost of a lot of network traffic ; or the model only exists in the form of a wavefront in GPU memory, above and below being offloaded to CPU

## Pitch
Test implementations and collect feedback

## Alternatives
Use Pipe, Megatron,..

**Comment by Young768** [comment]
Hi, is there any further plan regarding this feature? Im also looking into how to train very big model. I'd like to contribute sth on this if possible.

**Comment by blefaudeux** [comment]
@Young768 I've had little time to work on that recently, the attached code is a rough draft but the idea works, it has a lot of constraints on the model formulation as is (needs to be Sequential and fine grained enough, and will not work with all inputs). [Pipe](https://fairscale.readthedocs.io/en/latest/api/nn/pipe.html) is an option right now which should be more mature, else I'll do my best to advance on this one, and we have some other approaches incoming in FairScale in the near future

**Comment by Young768** [comment]
are you going to support intra-layer parallel for FairScale?

**Comment by blefaudeux** [comment]
> are you going to support intra-layer parallel for FairScale?

- yes, in the near future, no precise date to announce right now unfortunately
- not with this PR, the idea here is to offload parameters to and from CPU space, or at least experiment with the idea while sticking to pytorch

**Comment by blefaudeux** [comment]
Done by the team with FSDP
